server/tso.go

```diff
@@ -122,29 +123,50 @@ func (s *Server) updateTimestamp() error {
 		log.Warnf("clock offset: %v, prev: %v, now: %v", since, prev, now)
 		tsoCounter.WithLabelValues("slow_save").Inc()
 	}
-	// Avoid the same physical time stamp
-	if since <= updateTimestampGuard {
-		tsoCounter.WithLabelValues("skip_save").Inc()
-		log.Warnf("invalid physical timestamp, prev: %v, now: %v, re-update later", prev, now)
-		return nil
+
+	current := &atomicObject{
+		physical: now,
 	}
 
-	if subTimeByWallClock(now, s.lastSavedTime) >= 0 {
-		last := s.lastSavedTime
-		save := now.Add(s.cfg.TsoSaveInterval.Duration)
-		if err := s.saveTimestamp(save); err != nil {
-			return errors.Trace(err)
+	saved := s.lastSavedTime.Load().(time.Time)
+	// This is a normal situation.
+	if since > 0 {
+		// Avoid the same physical timestamp.
+		if since <= updateTimestampGuard {
+			tsoCounter.WithLabelValues("skip_save").Inc()
+			log.Warnf("invalid physical timestamp, prev: %v, now: %v, re-update later", prev, now)
+			return nil
 		}
-
-		tsoCounter.WithLabelValues("save_ok").Inc()
-		log.Debugf("save timestamp ok: prev %v last %v save %v", prev, last, save)
+		// If the time window is no more than updateTimestampGuard,
+		// it will adjust the time window.
+		if subTimeByWallClock(saved, now) <= updateTimestampGuard {
```

Comment:
better also validate the `TsoSaveInterval` large than `updateTimestampGuard` in config.

Reply:
They are not in the same order of magnitude, and thus it doesn't make much sense. What do you think?